swig/ChunkedArray.hpp

```diff
@@ -0,0 +1,240 @@
+ * track of number of inserted elements, but is not thread-safe (ok as usually input is a streaming iterator).
+ * For parallel input sources the low-level API must be used.
+ *
+ * Note: When using this for `LGBM_DatasetCreateFromMats` use a
+ *       chunk_size multiple of #num_cols for your dataset, so each chunk
+ *       contains "complete" instances.
+ *
+ * === High-level insert API intro ===
+ *
+ * The easiest way to use is:
+ *  0. ChunkedArray(chunk_size)  # Choose appropriate size
+ *  1. add(value)                # as many times as you want (will generate chunks as needed)
+ *  2. data() or void_data()     # retrieves a T** or void** pointer (useful for `LGBM_DatasetCreateFromMats`).
+ *
+ * Useful query methods (all O(1)):
+ *  - get_add_count()   # total count of added elements
+ *  - get_chunks_count()  # how many chunks are currently allocated.
+ *  - get_current_chunk_added_count()  # for the last add() chunk, how many items there are.
+ *  - get_chunk_size()    # Get constant chunk_size from constructor call.
+ *
+ *  With those you can generate int32_t sizes[]. Last chunk can be smaller than chunk_size, so, for any i:
+ *    - sizes[i<last] = get_chunk_size()
+ *    - sizes[i==last] = get_add_count()
+ *
+ *
+ * === Low-level insert API intro ===
+ *
+ * For advanced usage - useful for inserting in parallel - one can also:
+ *  1. call new_chunk() at any time for as many chunks as needed.  (thread-UNsafe)
+ *  2. call setitem(chunk, idx, value) to insert each value.       (thread-safe)
+ *
+ */
+template <class T>
+class ChunkedArray
+{
+  public:
+    ChunkedArray(size_t chunk_size)
+      : _chunk_size(chunk_size), _current_chunks_idx(0), _current_chunk_idx(0)
+    {
+       new_chunk();
+    }
+
+    ~ChunkedArray()
+    {
+        release();
+    }
+
+    /**
+     * Adds a value to the chunks sequentially.
+     * If the last chunk is full it creates a new one and appends to it.
+     */
+    void add(T value) {
+        if (! within_bounds(_current_chunks_idx, _current_chunk_idx)) {
+            new_chunk();
+            _current_chunks_idx += 1;
+            _current_chunk_idx = 0;
+        }
+
+        assert (setitem(_current_chunks_idx, _current_chunk_idx, value) == 0);
+        _current_chunk_idx += 1;
+    }
+
+    size_t get_add_count() const {
+        return _current_chunks_idx * _chunk_size + _current_chunk_idx;
+    }
+
+    size_t get_chunks_count() const {
+        return _chunks.size();
+    }
+
+    size_t get_current_chunk_added_count() const {
+        return _current_chunk_idx;
+    }
+
+    size_t get_chunk_size() const {
+        return _chunk_size;
+    }
+
+    /**
+     * Returns the pointer to the raw chunks data.
+     *
+     * @return T** pointer to raw data.
+     */
+    T **data() noexcept
+    {
+        return _chunks.data();
+    }
+
+    /**
+     * Returns the pointer to the raw chunks data, but cast to void**.
+     * This is so ``LGBM_DatasetCreateFromMats`` accepts it.
```

Comment:
wait, so no change to LGBM_DatasetCreateFromMats is required then?!  very cool if true.  I don't see any changes to LGBM_DatasetCreateFromMats in this PR.

Reply:
Yes, this is just the "glue code" to accumulate data from an outside stream of data into a regular C++ array of arrays which can be used by `LGBM_DatasetCreateFromMats`.